### DISCUSSION TITLE
requester在调用service的时候如何动态选择一个group?

### DISCUSSION THREAD

**Discussion by bearxu83** [alibaba/alibaba-rsocket-broker#141] [Q&A] [answered]
你好,我是阿里翱象isv的开发. 使用中遇到一个问题.

比如, 将同一个userServiceImpl部署到两台前置机groupA和groupB上, 在调用userService.findID()的时候,我们希望在前置机groupA上执行, groupB不执行,请问如何实现? 

我尝试了rsocketRequester.route, 但是

>   @Bean
>     public RSocketRequester rsocketRequester(UpstreamManager upstreamManager) {
>         LoadBalancedRSocket loadBalancedRSocket = upstreamManager.findBroker().getLoadBalancedRSocket();

报了loadBalancedRSocket 为空.

我又尝试了在调用的时候生成一个RSocketRemoteServiceBuilder. 

>         final RSocketRemoteServiceBuilder<SomeService> someServiceBuilder = RSocketRemoteServiceBuilder
>                 .client(SomeService.class)
>                 .group("groupA")
>                 .upstreamManager(upstreamManager);
>         log.info("start");
>         final SomeService someService = someServiceBuilder.build();
>         log.info("end1");
>         log.info("{}", someService.getSkuList());
>         log.info("end2");
>         return someService.getSkuList();

但在调用someService.getSkuList()的时候hang住了. 

请问Request在调请求的时候, 如何去选择endpoint或者group?谢谢

感觉第二种方法是对的, 可能upstreamManager是autowired的, 没有正确初始化?

**Comment by bearxu83** [comment]
在

> LoadBalancedRSocket loadBalancedRSocket = upstreamManager.findBroker().getLoadBalancedRSocket();

之前加一行

> upstreamManager.findBroker().init();

可以用了. 不知道有没有更好的方法.

**Reply by bearxu83** [reply]
使用rsocketRequester.route().retrieveMono(String.class)时, 返回一个Mono<String>的话,得到的字符串会比实际的多1个或2个字符在字符串开头的位置,比如希望得到"xxx", 实际拿到"0 xxx", 或者"\u0007xxx". 
如果不使用rsocketRequester.route, 这个场景是不是有官方推荐的方法?谢谢. @linux-china

**Comment by linux-china** [comment, marked as answer]
忘记贴这个啦  RSocket路由规则    https://github.com/alibaba/alibaba-rsocket-broker/wiki/RSocket-Routing    group解决不同的分组问题，如大家都懂的VIP和common分组，如果是开发和调试的时候，可以使用endpoint，如 `ip:192.168.1.33` 这样。  如果要动态的话，可能你需要创建几个bean，然后指定不同的名称，在调用的时候自己要进行对应的bean选择。
